### PR TITLE
Enable p-refinement via an Event

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

This PR will allow p-refinement to be done as an Event.
This requires the DG scheme to reactively check if the communicated mortar data is on the correct Mesh.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp
Comment (nilsdeppe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp]: Combine lines

File: src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp]: move to cpp

File: src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp]: move to cpp

File: src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp:70]: Do you need a `pup` function?

File: src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp]: Should we choose p-refine if we get split? And p-coarsen if we get Join? Or just ignore for now? It might be good to have an ERROR since this would indicate a decision is being ignored.

File: src/Evolution/DiscontinuousGalerkin/BoundaryData.hpp
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/BoundaryData.hpp]: Renumber the later entries.  (The numbering being wrong actually confuses doxygen, which surprises me.)

File: src/Evolution/DiscontinuousGalerkin/InboxTags.hpp
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/InboxTags.hpp]: Again, fix numbering.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp]: [optional] formatting

File: src/ParallelAlgorithms/Amr/Events/Events.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/Amr/Events/Events.hpp]: Sspelling.

File: src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp:507]: Call `boundary_data_history->at(mortar_id).clear_coupling_cache()`.

File: src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp:538]: Again, clear the cache, unless this is only modifying the entry that was just inserted, in which case add a comment that you don't have to clear the cache.

File: src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp]: Better message?  Probably print out the values.

File: src/ParallelAlgorithms/Amr/Events/Events.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/Amr/Events/Events.hpp:8]: Move reformatting to the commit where it was added.  Also remove the same-line closing comment.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Actions/Test_ApplyBoundaryCorrections.cpp
Comment (wthrowe) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Actions/Test_ApplyBoundaryCorrections.cpp]: Move to appropriate commit.

**Review by wthrowe — CHANGES_REQUESTED** [review]
clang-tidy errors are not in our code and probably wrong, so nothing to do for those.

File: src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp]: A bit more info would be helpful :)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp
Comment (knelli2) [inline src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp:97]: Should these changes go in a different commit?

**Review by knelli2 — COMMENTED** [review]
No need to wait for my approval

**Review by nilsdeppe — DISMISSED** [review]
My requests have been addressed. I don't need to look at this again before it's merged.

**Comment by wthrowe** [comment]
Still missing the LTS cache management.  Have you tested this in an LTS evolution?

**Review by wthrowe — APPROVED** [review]

**Comment by wthrowe** [comment]
We discussed some methods of testing this elsewhere.  I assume you've done some of those?

**Comment by wthrowe** [comment]
Discussed elsewhere.  The tests have been done.

**Pull request merged**
